crates/bevy_ecs/src/schedule/state.rs

```diff
@@ -40,24 +46,160 @@ pub use bevy_ecs_macros::States;
 /// }
 ///
 /// ```
-pub trait States: 'static + Send + Sync + Clone + PartialEq + Eq + Hash + Debug + Default {
-    /// The type returned when iterating over all [`variants`](States::variants) of this type.
-    type Iter: Iterator<Item = Self>;
+pub trait States: 'static + Send + Sync + Clone + PartialEq + Eq + Hash + Debug + Default {}
+
+/// Types that can match world-wide states.
+pub trait StateMatcher<S: States>: 'static + Send + Sync + Clone {
+    /// Check whether to match with the current state
+    fn match_state(&self, state: &S) -> bool;
+}
+
+impl<S: States> StateMatcher<S> for S {
+    fn match_state(&self, state: &S) -> bool {
+        self == state
+    }
+}
+
+/// A struct for containing a state matcher function
```

Comment:
This doc string isn't helpful: we're just restating the struct name :)

Explaining the `Fn` type would be particularly helpful to readers: tat's often quite challenging.

Reply:
Yes! Will do